### PR TITLE
Copy sim state

### PR DESCRIPTION
**Pull request by havahol**
New features:
- Simulator.copyState(otherSim) -> the simulator copies the eta, hu and hv field from otherSim, as well as its wind object and drifter positions.
Since this is intended to be used in ensemble reinitialization, all (possibly) perturbed variables should be copied in this function.

Behind the scenes:
The Simulator constructor is now given some lines of code that was the same across all sims.
The task about unifying the Simulator constructor signatures are still on the To-do-board.

### CODE REVIEW

**Comment by havahol** [comment]
[Implement device-to-device copy of buffers, and make it available to simulators so that they can be reinitialized with the ocean state of another simulator object.](https://trello.com/c/Ujy6NUMu/54-implement-device-to-device-copy-of-buffers-and-make-it-available-to-simulators-so-that-they-can-be-reinitialized-with-the-ocean)

File: python/SWESimulators/Simulator.py [hunk 1]
```diff
@@ -89,6 +167,38 @@ def downloadPrevTimestep(self):
...
+        self.cl_data.hv1.copyBuffer(self.cl_queue, otherSim.cl_data.hv1)
+        
+        # Question: Which parameters should we require equal, and which 
+        # should become equal?
+        self.wind_stress = otherSim.wind_stress
+        
+        self.drifters.setParticlePositions(otherSim.drifters.getParticlePositions())
+        self.drifters.setObservationPosition(otherSim.drifters.getObservationPosition())
+        
+        #self.drifters = None
```
Comment (setmar) [inline python/SWESimulators/Simulator.py]: Out-commented code :)

File: python/SWESimulators/Simulator.py [hunk 2]
```diff
@@ -89,6 +167,38 @@ def downloadPrevTimestep(self):
         """
         return self.cl_data.downloadPrevTimestep(self.cl_queue)
         
+    def copyState(self, otherSim):
+        """
+        Copies the state ocean state (eta, hu, hv) from the other simulator.
+        
+        This function is exposed to enable efficient re-initialization of
+        resampled ocean states. This means that all parameters which can be 
+        initialized/assigned a perturbation should be copied here as well.
```
Comment (setmar) [inline python/SWESimulators/Simulator.py]: Is wind part of the state? Why not bathymetry? (We might do perturbations on bathymetry at some stage.) Maybe we should write it as "state and wind"?

File: python/SWESimulators/CTCS.py [hunk 3]
```diff
@@ -73,61 +73,55 @@ def __init__(self, \
...
+                                   offset_x, offset_y, \
+                                   block_width, block_height)
+            
+        #Get kernels
+        self.u_kernel = Common.get_kernel(self.cl_ctx, "CTCS_U_kernel.opencl", block_width, block_height)
+        self.v_kernel = Common.get_kernel(self.cl_ctx, "CTCS_V_kernel.opencl", block_width, block_height)
+        self.eta_kernel = Common.get_kernel(self.cl_ctx, "CTCS_eta_kernel.opencl", block_width, block_height)
 
-              
+        #self.u_kernel_exec = self.u_kernel.computeUKernel
```
Comment (setmar) [inline python/SWESimulators/CTCS.py]: Out-commented code should be removed.

File: python/SWESimulators/Simulator.py (hunk 2 above)
Comment (havahol) [reply python/SWESimulators/Simulator.py]: You illustrate my point: "We **might** do perturbations on bathymetry". We might also perturb other things! So my suggestion is that when we perturb anything new, we should never expect that this parameter is handled specially before. As a consequence: Bathymetry should therefore be added in this function when we actually perturb it.

**Pull request merged**